<a href="https://colab.research.google.com/github/ShubhamCJagtap/Capstone_Project_1_EDA/blob/master/Capstone_project_1_Mobile_Events_EDA_colab_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 ## EDA steps to follow
 
**1.** [**Introduction**](#Section1)<br>
**2.** [**Problem Statement**](#Section2)<br>
**3.** [**Installing & Importing Libraries**](#Section3)<br>
  - **3.1** [**Installing Libraries**](#Section31)
  - **3.2** [**Upgrading Libraries**](#Section32)
  - **3.3** [**Importing Libraries**](#Section33)

**4.** [**Data Acquisition & Description**](#Section4)<br>
  - **4.1** [**Data Description**](#Section41)
  - **4.2** [**Data Information**](#Section42)

**5.** [**Data Pre-profiling**](#Section5)<br>
**6.** [**Data Cleaning**](#Section6)<br>
**7.** [**Data Post-profiling**](#Section7)<br>
**8.** [**Exploratory Data Analysis**](#Section8)<br>
  - **8.1** [**Outlet Metrics-based Analysis**](#Section81)
  - **8.2** [**Nutritional Value-based Analysis**](#Section82)
  - **8.3** [**Geographical Information-based Analysis**](#Section83)
  - **8.4** [**Outlet Metrics-based & Nutritional Value-based Analysis**](#Section84)
  - **8.5** [**Outlet Metrics-based & Geographical Information-based Analysis**](#Section85)
  - **8.6** [**Nutritional Value-based & Geographical Information-based Analysis**](#Section86)
  - **8.7** [**Outlet Metric-based, Menu Items & Geographical Information-based Analysis**](#Section87)

**9.** [**Summarization**](#Section9)<br>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Introduction

### Company Introduction
Your company for this project is InsaidTelecom.<br>

InsaidTelecom, one of the leading telecom players, understands that customizing offerings is very important for its business to stay competitive.<br>
Currently, InsaidTelecom is seeking to leverage behavioural data from more than 60% of the 50 million mobile devices active daily in India<br>
They are doing this to help their clients better understand and interact with their audiences.<br>


# problem Statement

## Current Scenario
In this consulting assignment, Insaidians are expected to build a dashboard.<br>
This dashboard will help us to understand a user's demographic characteristics based on their mobile usage, geolocation, and mobile device properties<br>
Doing so will help millions of developers and brand advertisers around the world pursue data-driven marketing efforts.
These efforts are relevant to their users and cater to their preferences.<br>

### Expected output

- EDA observations
- Dashboard

# Role

## Your Role
- To help the customer the consultants are expected to have depth of clarity in the underlying data.
- How much effort has been put into cleansing and purifying the data will decide how closely have you looked at the data..
- How detailed is the observation stated in the submission report and finally, how well a group presents their consulting journey.
- Please remember that this is an analytics consulting firm. Hence, your efforts in terms of finding user behaviour are going to directly impact the company's offerings.
- Do help the company understand what is the right way forward and suggest actionable insights from marketing and product terms.


### State to focus
- Delhi

### Provided Instructions

#### Dataset Description
- In this assignment, you are going to study the demographics of a user (gender and age) based on their app download and usage behaviours.
- The Data is collected from mobile apps that use Inside Telecom services. Full recognition and consent from an individual user of those apps have been obtained,and appropriate anonymization has been performed to protect privacy. Due to confidentiality, we won't provide details on how the gender and age data were obtained.
- Please treat them as accurate ground truth for prediction.


## Week 1 Task

- Fetch the data on to the notebook and ponder upon the challenges present in the data
- All observations supposed to be listed in observation sheet

# Import Libraries

In [2]:
!pip install mysql.connector

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 11.9 MB 5.2 MB/s 
  Created wheel for mysql.connector: filename=mysql_connector-2.2.9-cp37-cp37m-linux_x86_64.whl size=247965 sha256=17f96a8a50aaba959e522f1951072c3d6a765c997ff49e8e69650d3be0bf938d
  Stored in directory: /root/.cache/pip/wheels/42/2f/c3/692fc7fc1f0d8c06b9175d94f0fc30f4f92348f5df5af1b8b7
Successfully built mysql.connector


In [3]:
import pandas as pd
import numpy as np
import mysql.connector as connection # install !pip install mysql.connector
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)
pd.set_option('display.max_colwidth',None)
pd.set_option('mode.chained_assignment',None)

In [ ]:
Google_drive_link = 'https://drive.google.com/file/d/1Ir3rW0YTKmk7MSjVjCU_UGMQevhe1v9W/view'

# Collect Events Data

In [4]:
df = pd.read_csv('/content/drive/MyDrive/Capstone 1 Insaid/Database/events_data.csv')   #read csv file

In [5]:
df.head(10)

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968 -3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   
5   1078723 -5.124242e+17  2016-05-02 02:21:20  83.398244  17.768149   
6   2948755  5.536513e+18  2016-05-03 16:20:14  80.359796  13.128174   
7    280014 -8.879644e+18  2016-05-05 13:06:01  78.155397  16.390327   
8   2309828  3.020769e+18  2016-05-04 09:24:50  77.263922  28.757914   
9    720049  4.928556e+18  2016-05-06 19:00:52  72.844795  18.974929   

            city          state  
0          Delhi          Delhi  
1       Calcutta     WestBengal  
2          Delhi          Delhi  
3        Chennai      TamilNadu  
4         Bokaro      Jharkhand  
5  Visakhapatnam  AndhraPradesh  
6        Chennai      TamilNadu  
7       Wanparti  AndhraPradesh  
8          Delhi          Delhi  
9         Mumbai    Maharashtra

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype  
---  ------     -----  
 0   event_id   int64  
 1   device_id  float64
 2   timestamp  object 
 3   longitude  float64
 4   latitude   float64
 5   city       object 
 6   state      object 
dtypes: float64(3), int64(1), object(3)
memory usage: 173.7+ MB


## Observations (data.info())

- Total Entries  3252950
- timestamp type object ## need to change in datetime dtype

In [7]:
df.describe()

event_id     device_id     longitude      latitude
count  3.252950e+06  3.252497e+06  3.252527e+06  3.252527e+06
mean   1.626476e+06  1.012200e+17  7.815868e+01  2.168851e+01
std    9.390459e+05  5.316758e+18  4.235570e+00  5.789111e+00
min    1.000000e+00 -9.222957e+18  1.256700e+01  8.190110e+00
25%    8.132382e+05 -4.540611e+18  7.583543e+01  1.780171e+01
50%    1.626476e+06  1.726820e+17  7.726814e+01  2.216454e+01
75%    2.439713e+06  4.861813e+18  8.031916e+01  2.868278e+01
max    3.252950e+06  9.222849e+18  9.545946e+01  4.187190e+01

In [8]:
describe_df = df.describe()

In [9]:
describe_df[describe_df.index=='count']

event_id  device_id  longitude   latitude
count  3252950.0  3252497.0  3252527.0  3252527.0

In [10]:
describe_df

event_id     device_id     longitude      latitude
count  3.252950e+06  3.252497e+06  3.252527e+06  3.252527e+06
mean   1.626476e+06  1.012200e+17  7.815868e+01  2.168851e+01
std    9.390459e+05  5.316758e+18  4.235570e+00  5.789111e+00
min    1.000000e+00 -9.222957e+18  1.256700e+01  8.190110e+00
25%    8.132382e+05 -4.540611e+18  7.583543e+01  1.780171e+01
50%    1.626476e+06  1.726820e+17  7.726814e+01  2.216454e+01
75%    2.439713e+06  4.861813e+18  8.031916e+01  2.868278e+01
max    3.252950e+06  9.222849e+18  9.545946e+01  4.187190e+01

In [5]:
a = describe_df[describe_df.index=='count']['event_id']
for i in describe_df[describe_df.index=='count'].columns:
    print("Missing Values in {} are {}".format(i,a - describe_df[describe_df.index=='count'][i]))

NameError: ignored

## Observations (data.describe()

- Missing values in device_id = 453
- Missing values in longitude = 423
- Missing values in latitude = 423

## Analysis of object data state city

In [12]:
df[['state','city']].nunique()

state     32
city     933
dtype: int64

## Observations
Unique states present - 32<br>
Unique City present - 933

In [13]:
df[['state','city']].isna().sum()

state    377
city       0
dtype: int64

In [14]:
df['state'].value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
Meghalaya                     1515
AndamanandNicobarIslands      1515
HimachalPradesh               1409
Manipur             

Missing values present in state column - 377

In [7]:
#counting the unique values
for columns in df:
  print ("The unique entries in {} is {}". format (columns, df[columns].nunique()))


The unique entries in event_id is 3252950
The unique entries in device_id is 60865
The unique entries in timestamp is 588126
The unique entries in longitude is 60095
The unique entries in latitude is 60177
The unique entries in city is 933
The unique entries in state is 32


In [19]:
#counting the null values
for columns in df:
  print("The null values in {} is {}".format(columns, df[columns].isnull().sum()))


The null values in event_id is 0
The null values in device_id is 453
The null values in timestamp is 0
The null values in longitude is 0
The null values in latitude is 0
The null values in city is 0
The null values in state is 0


## Fill mssing values

## State

In [9]:
cities_list = (df[df['state'].isna()]['city'].unique()).tolist()
print(cities_list)

['Pune', 'Visakhapatnam', 'Indore', 'Chennai', 'Delhi', 'Channapatna', 'Jaipur', 'Gangarampur', 'Arambagh']


In [10]:
state_list = []
for i in cities_list:
  print(i,df[(df['city']==i)&(df['state'].isna()==False)]['state'].unique())
  state_list.append(df[(df['city']==i)&(df['state'].isna()==False)]['state'].unique())

Pune ['Maharashtra']
Visakhapatnam ['AndhraPradesh']
Indore ['MadhyaPradesh']
Chennai ['TamilNadu']
Delhi ['Delhi']
Channapatna ['Karnataka']
Jaipur ['Rajasthan']
Gangarampur ['WestBengal']
Arambagh ['WestBengal']


In [11]:
df.loc[(df['city']=='Pune')&(df['state'].isna()),'state']='Maharashtra'
df.loc[(df['city']=='Visakhapatnam')&(df['state'].isna()),'state']='AndhraPradesh'
df.loc[(df['city']=='Indore')&(df['state'].isna()),'state']='MadhyaPradesh'
df.loc[(df['city']=='Chennai')&(df['state'].isna()),'state']='TamilNadu'
df.loc[(df['city']=='Delhi')&(df['state'].isna()),'state']='Delhi'
df.loc[(df['city']=='Channapatna')&(df['state'].isna()),'state']='Karnataka'
df.loc[(df['city']=='Jaipur')&(df['state'].isna()),'state']='Rajasthan'
df.loc[(df['city']=='Gangarampur')&(df['state'].isna()),'state']='WestBengal'
df.loc[(df['city']=='Arambagh')&(df['state'].isna()),'state']='WestBengal'

## Latitude and longitude

In [13]:
device_id_lat_lon_missing = df[(df['longitude'].isna())&(df['latitude'].isna())]['device_id'].unique()
print(device_id_lat_lon_missing)

[ 1.77963102e+18 -8.79056003e+18  4.86560657e+18  7.98387143e+18
 -7.36484063e+18  5.37559902e+18  5.51865960e+18  4.58017944e+18
 -6.03379839e+18  7.66213914e+18  2.35097940e+18  4.37503983e+18
  4.78258205e+18  1.32005098e+18  5.73136927e+18  3.73828049e+18
  3.05848775e+17  7.01860038e+18  5.11518528e+18  6.77407134e+18
  3.09916855e+18]


In [14]:
lst_id_lat_lon = []
for i in device_id_lat_lon_missing:
  a = df[(df['longitude'].isna()==False)&(df['latitude'].isna()==False)&(df['device_id']==i)][['device_id','longitude','latitude']].value_counts()
  lst_id_lat_lon.append(a.index.values)

In [15]:
lst_id_lat_lon

[array([(1.7796310234394053e+18, 77.235416, 28.738901)], dtype=object),
 array([(-8.790560034584249e+18, 75.933778, 22.761965)], dtype=object),
 array([(4.865606571395669e+18, 83.35144199999998, 17.772251999999995)],
       dtype=object),
 array([(7.983871429718485e+18, 80.33543499999998, 13.189053)],
       dtype=object),
 array([(-7.364840628879981e+18, 73.931944, 18.54074)], dtype=object),
 array([(5.375599021847303e+18, 77.283447, 28.718392)], dtype=object),
 array([(5.518659601894558e+18, 83.384242, 17.817945)], dtype=object),
 array([(4.580179444377398e+18, 75.835756, 26.960220000000003)],
       dtype=object),
 array([(-6.033798388179802e+18, 83.31978699999998, 17.816579)],
       dtype=object),
 array([(7.66213914371992e+18, 80.30927199999998, 13.149176)], dtype=object),
 array([(2.3509794022024387e+18, 75.888474, 22.745063)], dtype=object),
 array([(4.37503983023287e+18, 75.824489, 26.980345000000003)],
       dtype=object),
 array([(4.782582047729166e+18, 77.22522, 28.718379)

In [16]:
id = []
lat = []
lon = []
for j in lst_id_lat_lon:
  id.append(j[0][0])
  lon.append(j[0][1])
  lat.append(j[0][2])

In [17]:
for i in range(0,len(id)):
  print("id = {}, lat ={}, lon ={}".format(id[i],lat[i],lon[i]))

id = 1.7796310234394053e+18, lat =28.738901, lon =77.235416
id = -8.790560034584249e+18, lat =22.761965, lon =75.933778
id = 4.865606571395669e+18, lat =17.772251999999995, lon =83.35144199999998
id = 7.983871429718485e+18, lat =13.189053, lon =80.33543499999998
id = -7.364840628879981e+18, lat =18.54074, lon =73.931944
id = 5.375599021847303e+18, lat =28.718392, lon =77.283447
id = 5.518659601894558e+18, lat =17.817945, lon =83.384242
id = 4.580179444377398e+18, lat =26.960220000000003, lon =75.835756
id = -6.033798388179802e+18, lat =17.816579, lon =83.31978699999998
id = 7.66213914371992e+18, lat =13.149176, lon =80.30927199999998
id = 2.3509794022024387e+18, lat =22.745063, lon =75.888474
id = 4.37503983023287e+18, lat =26.980345000000003, lon =75.824489
id = 4.782582047729166e+18, lat =28.718379, lon =77.22522
id = 1.3200509770197112e+18, lat =26.211919, lon =87.570736
id = 5.731369272434022e+18, lat =13.153332, lon =80.34361299999998
id = 3.738280493453803e+18, lat =22.76201, lon

In [18]:
for i in range(0,len(id)):
  df.loc[(df['device_id']==id[i])&(np.isnan(df['latitude'])),'latitude'] = lat[i]
  df.loc[(df['device_id']==id[i])&(np.isnan(df['longitude'])),'longitude'] = lon[i]

## Device Id

In [20]:
device_id_missing_lon = df[df['device_id'].isna()]['longitude'].unique()
device_id_missing_lat = df[df['device_id'].isna()]['latitude'].unique()

In [21]:
device_id_missing_lon,device_id_missing_lat

(array([75.882956, 75.888487, 75.846007, 75.923332, 75.992551, 73.862756,
        73.860165, 83.357991, 77.292481, 73.926499, 80.343613, 80.309272,
        77.274814, 83.371738, 75.95805 , 83.342711, 75.836167, 80.335435,
        73.169345, 70.686387, 77.303153]),
 array([22.814519, 26.948689, 26.960796, 22.777781, 31.561747, 18.628057,
        18.566925, 17.805195, 28.719966, 18.614812, 13.153332, 13.149176,
        28.721053, 17.752819, 22.817526, 17.822906, 26.95399 , 13.189053,
        21.194283, 21.790693, 28.728888]))

In [22]:
lst_id_lon = []
for i in device_id_missing_lon:
  a = df[(df['device_id'].isna()==False)&(df['longitude']==i)][['device_id','longitude','latitude']].value_counts()
  lst_id_lon.append(a.index.values)

In [23]:
lst_id_lat = []
for i in device_id_missing_lat:
  a = df[(df['device_id'].isna()==False)&(df['latitude']==i)][['device_id','longitude','latitude']].value_counts()
  lst_id_lat.append(a.index.values)

In [24]:
# a = 0
# for i in range(0,len(lst_id_lon)):
#   if lst_id_lon[i] in lst_id_lat:
#     pass
#   else:
#     a=1
#     print("{} and {} are different at index {}".format(lst_id_lon[i],lst_id_lat[i],i))
# if a==0:
#   print("both are same")

print(lst_id_lat)
print(lst_id_lon)

[array([(3.1328613559497626e+18, 75.88295600000002, 22.814519)],
      dtype=object), array([(-9.177251100412943e+17, 75.888487, 26.948689)], dtype=object), array([(-8.460337302442786e+18, 75.846007, 26.960796)], dtype=object), array([(7.597735768595925e+18, 75.92333199999999, 22.777781)],
      dtype=object), array([(1.750778632182067e+18, 75.99255099999998, 31.561747)],
      dtype=object), array([(9.027085705222633e+18, 73.862756, 18.628057)], dtype=object), array([(8.539870903150472e+18, 73.86016500000002, 18.566925)],
      dtype=object), array([(6.844192269280791e+18, 83.357991, 17.805195)], dtype=object), array([(5.098778421671838e+18, 77.29248100000002, 28.719966000000003)],
      dtype=object), array([(1.771585711650787e+17, 73.92649899999998, 18.614812)],
      dtype=object), array([(5.731369272434022e+18, 80.34361299999998, 13.153332)],
      dtype=object), array([(7.66213914371992e+18, 80.30927199999998, 13.149176)], dtype=object), array([(3.724654925765159e+18, 77.27481399

In [25]:
id_lat = []
lat_lat = []
lon_lat = []
for j in lst_id_lat:
  id_lat.append(j[0][0])
  lon_lat.append(j[0][1])
  lat_lat.append(j[0][2])

In [26]:
id_lon = []
lat_lon = []
lon_lon = []
for j in lst_id_lat:
  id_lon.append(j[0][0])
  lon_lon.append(j[0][1])
  lat_lon.append(j[0][2])

In [27]:
id_lat,id_lon

([3.1328613559497626e+18,
  -9.177251100412943e+17,
  -8.460337302442786e+18,
  7.597735768595925e+18,
  1.750778632182067e+18,
  9.027085705222633e+18,
  8.539870903150472e+18,
  6.844192269280791e+18,
  5.098778421671838e+18,
  1.771585711650787e+17,
  5.731369272434022e+18,
  7.66213914371992e+18,
  3.724654925765159e+18,
  -3.1309031770761733e+18,
  -2.399250255599833e+18,
  6.619714762103552e+18,
  3.562355872640618e+18,
  7.983871429718485e+18,
  -8.215770519233686e+18,
  -1.688015122502424e+18,
  -6.875585507485886e+18],
 [3.1328613559497626e+18,
  -9.177251100412943e+17,
  -8.460337302442786e+18,
  7.597735768595925e+18,
  1.750778632182067e+18,
  9.027085705222633e+18,
  8.539870903150472e+18,
  6.844192269280791e+18,
  5.098778421671838e+18,
  1.771585711650787e+17,
  5.731369272434022e+18,
  7.66213914371992e+18,
  3.724654925765159e+18,
  -3.1309031770761733e+18,
  -2.399250255599833e+18,
  6.619714762103552e+18,
  3.562355872640618e+18,
  7.983871429718485e+18,
  -8.215770

In [28]:
a = 0
for i in range(0,len(id_lon)):
  if id_lon[i]==id_lat[i]:
    pass
  else:
    a=1
    print("{} and {} are different at index {}".format(id_lon[i],id_lat[i],i))
if a==0:
  print("both are same")

both are same


In [29]:
for i in range(0,len(id_lat)):
  df.loc[(df['latitude']==lat_lat[i])&(np.isnan(df['device_id'])),'device_id'] = id_lat[i]

In [30]:
#counting the null values
for columns in df:
  print("The null values in {} is {}".format(columns, df[columns].isnull().sum()))

The null values in event_id is 0
The null values in device_id is 0
The null values in timestamp is 0
The null values in longitude is 0
The null values in latitude is 0
The null values in city is 0
The null values in state is 0


In [31]:
df.shape,df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype  
---  ------     -----  
 0   event_id   int64  
 1   device_id  float64
 2   timestamp  object 
 3   longitude  float64
 4   latitude   float64
 5   city       object 
 6   state      object 
dtypes: float64(3), int64(1), object(3)
memory usage: 173.7+ MB


((3252950, 7), None)

# Gender Age data from Mysql Server

In [ ]:
# try:
#     mydb = connection.connect(host="cpanel.insaid.co", database = 'Capstone1',user="student", passwd="student",use_pure=True)
#     query = "Select * from gender_age_train;"
#     result_dataFrame = pd.read_sql(query,mydb)
#     mydb.close() #close the connection
# except Exception as e:
#     mydb.close()
#     print(str(e))

In [ ]:
# result_dataFrame.to_csv('gender_age.csv')

In [ ]:
# df1 = pd.read_csv('gender_age.csv')

# Phone brand device model data from Mysql Server

In [ ]:
# try:
#     mydb = connection.connect(host="cpanel.insaid.co", database = 'Capstone1',user="student", passwd="student",use_pure=True)
#     query = "Select * from phone_brand_device_model;"
#     result_dataFrame = pd.read_sql(query,mydb)
#     mydb.close() #close the connection
# except Exception as e:
#     mydb.close()
#     print(str(e))

In [ ]:
# result_dataFrame.to_csv('phone_brand_device_model.csv')

In [ ]:
# df2 = pd.read_csv('phone_brand_device_model.csv')

In [ ]:
# for column in df.columns:
#   print(df[column].unique())
duplicated_list = df.duplicated()

In [ ]:
df[duplicated_list]

Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state]
Index: []

## Observations
No duplicate entries present in the dataset

# Same analysis for dataset 2 Gender_age.csv

In [32]:
df2 = pd.read_csv('https://raw.githubusercontent.com/ShubhamCJagtap/Capstone_Project_1_EDA/master/gender_age.csv',index_col=0)

In [76]:
df2.head()

device_id gender  age   group
0 -8076087639492063270      M   35  M32-38
1 -2897161552818060146      M   35  M32-38
2 -8260683887967679142      M   35  M32-38
3 -4938849341048082022      M   30  M29-31
4   245133531816851882      M   30  M29-31

In [73]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  74645 non-null  int64 
 1   device_id   74645 non-null  int64 
 2   gender      74645 non-null  object
 3   age         74645 non-null  int64 
 4   group       74645 non-null  object
dtypes: int64(3), object(2)
memory usage: 2.8+ MB


# Observations

Total Entries - 74646<br>
No Null values detected<br>
More analysis required for group column

In [74]:
df2.describe()

Unnamed: 0     device_id           age
count  74645.000000  7.464500e+04  74645.000000
mean   37322.000000 -7.491354e+14     31.410342
std    21548.299759  5.327150e+18      9.868735
min        0.000000 -9.223067e+18      1.000000
25%    18661.000000 -4.617367e+18     25.000000
50%    37322.000000 -1.841362e+16     29.000000
75%    55983.000000  4.636656e+18     36.000000
max    74644.000000  9.222849e+18     96.000000

In [ ]:
df2.isna().sum()

Unnamed: 0    0
device_id     0
gender        0
age           0
group         0
dtype: int64

## Observation gender_age
age minimum - 1 , maximum - 96 and mean = 31.41 , median = 29<br>
Mean and median don't have high difference <br>
Assumption check - age is normally distributed ??

In [ ]:
df2['gender'].value_counts()

M    47904
F    26741
Name: gender, dtype: int64

## Observation

male entries are 47904<br>
female entries are 26741<br>

In [ ]:
df2.nunique()

Unnamed: 0    74645
device_id     74645
gender            2
age              85
group            12
dtype: int64

In [ ]:
df2['group'].unique()

array(['M32-38', 'M29-31', 'F24-26', 'F33-42', 'F27-28', 'M39+', 'M23-26',
       'M27-28', 'M22-', 'F43+', 'F23-', 'F29-32'], dtype=object)

In [ ]:
print("Male above age 39 are {}".format(df2[(df2['age']>39)&(df2['gender']=='M')]['age'].sum()))

print("Female above age 43 are {}".format(df2[(df2['age']>43)&(df2['gender']=='F')]['age'].sum()))


Male above age 39 are 372416
Female above age 43 are 199865


In [ ]:
print("Male below age 22 are {}".format(df2[(df2['age']<22)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 23 are {}".format(df2[(df2['age']<23)&(df2['gender']=='F')]['age'].sum()))

Male below age 22 are 108781
Female below age 23 are 80571


In [ ]:
print("Male below age 10 are {}".format(df2[(df2['age']<10)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 10 are {}".format(df2[(df2['age']<10)&(df2['gender']=='F')]['age'].sum()))

Male below age 10 are 25
Female below age 10 are 0


In [ ]:
print("Male below age 5 are {}".format(df2[(df2['age']<5)&(df2['gender']=='M')]['age'].sum()))
print("Female below age 5 are {}".format(df2[(df2['age']<5)&(df2['gender']=='F')]['age'].sum()))

Male below age 5 are 1
Female below age 5 are 0


In [ ]:
df2.sort_values('age',ascending=True)[['age','group']].head(25)

age group
10385    1  M22-
60833    6  M22-
60832    6  M22-
60779    6  M22-
60831    6  M22-
57703   10  M22-
33595   10  M22-
15107   10  M22-
72936   10  F23-
72935   10  F23-
53296   11  F23-
53282   11  F23-
53994   11  F23-
52335   11  F23-
54609   12  M22-
53603   12  M22-
60527   12  M22-
61136   12  F23-
53642   12  F23-
59192   12  F23-
58421   12  M22-
61398   12  M22-
53136   12  F23-
55195   12  M22-
11510   12  M22-

In [ ]:
df2.sort_values('age',ascending=False)[['age','group']].head(25)

age group
5958    96  F43+
60872   94  F43+
3704    90  M39+
32290   89  F43+
2971    89  F43+
32291   89  F43+
3068    88  F43+
3560    88  F43+
19602   88  M39+
3065    88  F43+
3067    88  M39+
2972    87  M39+
49530   87  M39+
2973    87  M39+
43127   86  M39+
43128   86  M39+
60342   86  M39+
60349   86  F43+
27797   85  M39+
60348   85  M39+
37118   85  M39+
3573    84  M39+
2799    84  F43+
9150    83  M39+
9151    83  M39+

## identified Question
What should be the age range to be considered in EDA? 

# Unique Values
Gender - 2 (M,F)<br>
age - 85<br>
group - 12<br>

#groups are present as per gender

Male age group starts from M22- till M39+<br>
Female age group starts from F23- till F43+<br>

In [ ]:
duplicated_list_df2 = df2.drop('Unnamed: 0',axis=1).duplicated()

In [ ]:
df2.drop('Unnamed: 0',axis=1)[duplicated_list_df2]

Empty DataFrame
Columns: [device_id, gender, age, group]
Index: []

#Analysis of Phone_brand_dataset

In [33]:
df3 = pd.read_csv('https://raw.githubusercontent.com/ShubhamCJagtap/Capstone_Project_1_EDA/master/phone_brand_device_model.csv',index_col=0)

In [ ]:
df3.head(10)

Unnamed: 0            device_id phone_brand        device_model
0           0  1877775838486905855        vivo                 Y13
1           1 -3766087376657242966          小米                V183
2           2 -6238937574958215831        OPPO                 R7s
3           3  8973197758510677470          三星               A368t
4           4 -2015528097870762664          小米             红米Note2
5           5 -8896952452425303597          小米             红米Note3
6           6 -2634171391322875113          小米                S5.1
7           7 -2526386806972133373          三星  Galaxy Grand Prime
8           8  1427934937997164515          小米               MI 2S
9           9  2866484947180243791          三星       Galaxy Note 2

In [78]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   device_id     87726 non-null  int64 
 1   phone_brand   87726 non-null  object
 2   device_model  87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.7+ MB


In [ ]:
df3.isna().sum()

Unnamed: 0      0
device_id       0
phone_brand     0
device_model    0
dtype: int64

## Observation
Total entries - 87726<br>
No missing values

In [ ]:
df3.nunique()

Unnamed: 0      87726
device_id       87726
phone_brand       116
device_model     1467
dtype: int64

In [ ]:
df3['phone_brand'].unique()

array(['vivo', '小米', 'OPPO', '三星', '酷派', '联想 ', '华为', '奇酷', '魅族', '斐讯',
       '中国移动', 'HTC', '天语', '至尊宝', 'LG', '欧博信', '优米', 'ZUK', '努比亚', '惠普',
       '尼比鲁', '美图', '乡米', '摩托罗拉', '梦米', '锤子', '富可视', '乐视', '海信', '百立丰',
       '一加', '语信', '海尔', '酷比', '纽曼', '波导', '朵唯', '聆韵', 'TCL', '酷珀', '爱派尔',
       'LOGO', '青葱', '果米', '华硕', '昂达', '艾优尼', '康佳', '优购', '邦华', '赛博宇华',
       '黑米', 'Lovme', '先锋', 'E派', '神舟', '诺基亚', '普耐尔', '糖葫芦', '亿通', '欧新',
       '米奇', '酷比魔方', '蓝魔', '小杨树', '贝尔丰', '糯米', '米歌', 'E人E本', '西米', '大Q',
       '台电', '飞利浦', '唯米', '大显', '长虹', '维图', '青橙', '本为', '虾米', '夏新', '帷幄',
       '百加', 'SUGAR', '欧奇', '世纪星', '智镁', '欧比', '基伍', '飞秒', '德赛', '易派',
       '谷歌', '金星数码', '广信', '诺亚信', 'MIL', '白米', '大可乐', '宝捷讯', '优语', '首云',
       '瑞米', '瑞高', '沃普丰', '摩乐', '鲜米', '凯利通', '唯比', '欧沃', '丰米', '恒宇丰',
       '奥克斯', '西门子', '欧乐迪', 'PPTV'], dtype=object)

# Observation
phone_brand have many chinese brand names and company provided equivalent english names are limited.<br>


Brand Name	Brand English Mapping<br><br>
'华为'	'Huawei'<br>
'小米'	'Xiaomi'<br>
'三星'	'Samsung'<br>
'vivo'	'vivo'<br>
'OPPO'	'OPPO'<br>
'魅族'	'Meizu'<br>
'酷派'	'Coolpad'<br>
'乐视'	'LeEco'<br>
'联想 '	'Lenovo'<br>
'HTC'	'HTC'<br><br>

We shall check these brand names for Delhi only


In [ ]:
for i in df3['device_model'].unique():
  print(i,end='   ')

Y13   V183   R7s   A368t   红米Note2   红米Note3   S5.1   Galaxy Grand Prime   MI 2S   Galaxy Note 2   R1C   大神F1Plus   note顶配版   Galaxy Mega 2   星星1号   Galaxy S4   MI 3   Find 7   C8816   红米note   X6 L   Galaxy Win   青春版   MX3   Y35   G520-5000   荣耀4A   魅蓝Note 2   Y560-CL00   荣耀7i   3000   荣耀畅玩4C   C730Lw   红米2A   A1   One   Galaxy J7   R3   MX2   荣耀畅玩5   荣耀7   MI 4   One M8   红米1S   麦芒4   Galaxy S6 Edge   A3800d   P780   Ascend G7   荣耀6   Xperia SL   Y17T   荣耀畅玩4X   Ascend P7   Galaxy Grand 2   荣耀3X畅玩版   荣耀X2   Ascend P8   R2017   魅蓝NOTE   Mate 2   M8   T87+   X710L   A890e   R7   X5M   MX5   Galaxy Mega 5.8   L81   Xperia Z1   荣耀6 Plus   X5Max+   Galaxy J5   Galaxy Note 4   A31   GN706L   G3   P8 Lite   Sensation XE with Beats Audio Z715e   A880   荣耀+   X5L   8720L   IVO 6655   R7 Plus   荣耀畅玩4   MX4   Y635-TL00   Galaxy S3   X3L   3   R807   黄金斗士A8   One X   小米note   R6007   Galaxy S2 Plus   R1   荣耀3C   小米4C   R823T   Mate 7   MI 2   F2   Galaxy S5   Xperia TX   Galaxy Trend DUOS 2   Z1

## Observation
In device model, few names are in chinese. <br>
No equivalent english names provided

In [ ]:
duplicate_in_df3 = df3.drop('Unnamed: 0',axis=1).duplicated()

In [ ]:
df3.drop('Unnamed: 0',axis=1)[duplicate_in_df3]

Empty DataFrame
Columns: [device_id, phone_brand, device_model]
Index: []

No Duplicate values present

In [ ]:
# id = []
# miss = []
# for i in df2['device_id'].unique():
#   if i in df['device_id'].unique():
#     id.append(i)
#   else:
#     miss.append(i)

KeyboardInterrupt: ignored

Merge

In [34]:
df['device_id'] = df['device_id'].apply(lambda x: abs(x))
df2['device_id'] = df2['device_id'].apply(lambda x: abs(x))
df3['device_id'] = df3['device_id'].apply(lambda x: abs(x))

In [35]:
merged_1 = df.merge(df2, how = "left", on = "device_id")

In [36]:
merged_1.head()

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968  3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   

       city       state gender   age   group  
0     Delhi       Delhi      M  35.0  M32-38  
1  Calcutta  WestBengal    NaN   NaN     NaN  
2     Delhi       Delhi    NaN   NaN     NaN  
3   Chennai   TamilNadu    NaN   NaN     NaN  
4    Bokaro   Jharkhand    NaN   NaN     NaN

In [37]:
merged_1.shape

(3252950, 10)

In [38]:
merged_1.isna().sum()

event_id           0
device_id          0
timestamp          0
longitude          0
latitude           0
city               0
state              0
gender       3235968
age          3235968
group        3235968
dtype: int64

In [39]:
device_id_list_merged_1 = merged_1[(merged_1['gender'].isna()==False)&(merged_1['age'].isna()==False)&(merged_1['group'].isna()==False)]['device_id'].unique()

In [40]:
len(device_id_list_merged_1)

406

In [41]:
for i in device_id_list_merged_1[:10]:
  print(merged_1[(merged_1['device_id']==i)&(merged_1['gender'].isna())])

Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns: [event_id, device_id, timestamp, longitude, latitude, city, state, gender, age, group]
Index: []
Empty DataFrame
Columns:

In [42]:
merged_1.nunique()

event_id     3252950
device_id      60865
timestamp     588126
longitude      60095
latitude       60177
city             933
state             32
gender             2
age               51
group             12
dtype: int64

### first and last dataset

In [43]:
merged_2 = df.merge(df3, how = "left", on = "device_id")

In [44]:
merged_2.nunique()

event_id        3252950
device_id         60865
timestamp        588126
longitude         60095
latitude          60177
city                933
state                32
phone_brand          26
device_model        191
dtype: int64

In [45]:
merged_2.isna().sum()

event_id              0
device_id             0
timestamp             0
longitude             0
latitude              0
city                  0
state                 0
phone_brand     3235968
device_model    3235968
dtype: int64

In [46]:
device_id_list_merged_2 = merged_2[(merged_2['phone_brand'].isna()==False)&(merged_2['device_model'].isna()==False)]['device_id'].unique()

In [47]:
len(device_id_list_merged_2)

406

In [49]:
device_id_list_merged_2

array([2.97334779e+18, 1.44704917e+16, 3.87252179e+18, 4.97307639e+16,
       8.76587737e+18, 2.04981320e+16, 6.25452661e+15, 1.18022900e+18,
       2.72877884e+16, 2.68295479e+17, 4.05126807e+18, 7.28578686e+17,
       3.04562196e+18, 1.84136202e+16, 2.79778990e+16, 6.85719514e+15,
       2.24038934e+16, 1.71525482e+18, 1.52687161e+16, 4.54188019e+17,
       1.36785975e+18, 2.47071895e+17, 4.89144457e+17, 1.10568779e+18,
       6.71187700e+17, 9.94790294e+16, 4.89270408e+17, 2.66249525e+17,
       1.14999918e+18, 2.94044664e+15, 9.51773725e+17, 1.64535170e+17,
       1.32878992e+18, 1.19329363e+16, 1.16111271e+18, 8.22137105e+14,
       9.67151454e+17, 2.03614021e+17, 2.77799711e+16, 5.03873307e+15,
       2.87603673e+15, 1.04573805e+18, 5.90096778e+15, 2.56886687e+15,
       6.57709855e+16, 4.49698453e+15, 1.00683987e+16, 1.37006236e+17,
       3.47294291e+15, 1.53238293e+17, 5.02110504e+15, 7.40599713e+16,
       1.36421321e+15, 1.13818766e+18, 1.87007480e+18, 3.22537831e+16,
      

In [48]:
device_id_list_merged_2 == device_id_list_merged_1

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [50]:
sum = 0
for i in device_id_list_merged_1:
  print(df[df['device_id']==i]['device_id'].value_counts())
  sum = sum + df[df['device_id']==i]['device_id'].value_counts().values

2.973348e+18    194
Name: device_id, dtype: int64
1.447049e+16    49
Name: device_id, dtype: int64
3.872522e+18    69
Name: device_id, dtype: int64
4.973076e+16    333
Name: device_id, dtype: int64
8.765877e+18    466
Name: device_id, dtype: int64
2.049813e+16    27
Name: device_id, dtype: int64
6.254527e+15    334
Name: device_id, dtype: int64
1.180229e+18    60
Name: device_id, dtype: int64
2.728779e+16    1091
Name: device_id, dtype: int64
2.682955e+17    20
Name: device_id, dtype: int64
4.051268e+18    20
Name: device_id, dtype: int64
7.285787e+17    43
Name: device_id, dtype: int64
3.045622e+18    19
Name: device_id, dtype: int64
1.841362e+16    4
Name: device_id, dtype: int64
2.797790e+16    200
Name: device_id, dtype: int64
6.857195e+15    159
Name: device_id, dtype: int64
2.240389e+16    128
Name: device_id, dtype: int64
1.715255e+18    18
Name: device_id, dtype: int64
1.526872e+16    297
Name: device_id, dtype: int64
4.541880e+17    151
Name: device_id, dtype: int64
1.367860e+

In [44]:
sum

array([16982])

In [51]:
test_merge_1 = df.merge(df2, how = "inner", on = "device_id")
test_merge_2 = df.merge(df3, how = "inner", on = "device_id")

In [52]:
test_merge_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16982 entries, 0 to 16981
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   event_id      16982 non-null  int64  
 1   device_id     16982 non-null  float64
 2   timestamp     16982 non-null  object 
 3   longitude     16982 non-null  float64
 4   latitude      16982 non-null  float64
 5   city          16982 non-null  object 
 6   state         16982 non-null  object 
 7   phone_brand   16982 non-null  object 
 8   device_model  16982 non-null  object 
dtypes: float64(3), int64(1), object(5)
memory usage: 1.3+ MB


In [53]:
test_merge_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16982 entries, 0 to 16981
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   event_id   16982 non-null  int64  
 1   device_id  16982 non-null  float64
 2   timestamp  16982 non-null  object 
 3   longitude  16982 non-null  float64
 4   latitude   16982 non-null  float64
 5   city       16982 non-null  object 
 6   state      16982 non-null  object 
 7   gender     16982 non-null  object 
 8   age        16982 non-null  int64  
 9   group      16982 non-null  object 
dtypes: float64(3), int64(2), object(5)
memory usage: 1.4+ MB


In [54]:
test_merge_3 = test_merge_1.merge(test_merge_2[['device_id','phone_brand', 'device_model']], how = "left", on = "device_id")

In [55]:
test_merge_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3640706 entries, 0 to 3640705
Data columns (total 12 columns):
 #   Column        Dtype  
---  ------        -----  
 0   event_id      int64  
 1   device_id     float64
 2   timestamp     object 
 3   longitude     float64
 4   latitude      float64
 5   city          object 
 6   state         object 
 7   gender        object 
 8   age           int64  
 9   group         object 
 10  phone_brand   object 
 11  device_model  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 361.1+ MB


In [56]:
test_merge_3.nunique()

event_id        16982
device_id         406
timestamp       16553
longitude         406
latitude          406
city              197
state              26
gender              2
age                51
group              12
phone_brand        26
device_model      191
dtype: int64

In [57]:
test_merge_3.shape

(3640706, 12)

In [58]:
test_merge_3.isna().sum()

event_id        0
device_id       0
timestamp       0
longitude       0
latitude        0
city            0
state           0
gender          0
age             0
group           0
phone_brand     0
device_model    0
dtype: int64

In [59]:
duplicated_df = test_merge_3.duplicated()

In [60]:
duplicated_df.value_counts()

True     3623724
False      16982
dtype: int64

In [61]:
test_merge_3.drop_duplicates(inplace=True)

In [62]:
test_merge_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16982 entries, 0 to 3640705
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   event_id      16982 non-null  int64  
 1   device_id     16982 non-null  float64
 2   timestamp     16982 non-null  object 
 3   longitude     16982 non-null  float64
 4   latitude      16982 non-null  float64
 5   city          16982 non-null  object 
 6   state         16982 non-null  object 
 7   gender        16982 non-null  object 
 8   age           16982 non-null  int64  
 9   group         16982 non-null  object 
 10  phone_brand   16982 non-null  object 
 11  device_model  16982 non-null  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 1.7+ MB


In [63]:
test_merge_3.nunique()

event_id        16982
device_id         406
timestamp       16553
longitude         406
latitude          406
city              197
state              26
gender              2
age                51
group              12
phone_brand        26
device_model      191
dtype: int64

In [64]:
import calendar

test_merge_3['timestamp'] = pd.to_datetime(test_merge_3['timestamp'])
test_merge_3['year'] = pd.DatetimeIndex(test_merge_3['timestamp']).year
test_merge_3['month'] = pd.DatetimeIndex(test_merge_3['timestamp']).month

test_merge_3['month_name'] = test_merge_3['month'].apply(lambda x: calendar.month_abbr[x])
test_merge_3['day_of_week'] = test_merge_3['timestamp'].dt.day_name()
test_merge_3['hour'] = test_merge_3['timestamp'].dt.hour

In [65]:
dict = {'华为':'Huawei',
'小米':'Xiaomi',
'三星':'Samsung',
'vivo':'vivo',
'OPPO': 'OPPO',
'魅族': 'Meizu',
'酷派': 'Coolpad',
'乐视':'LeEco',
'联想 ': 'Lenovo',
'HTC': 'HTC'}

In [66]:
test_merge_3.replace({'phone_brand':dict},inplace=True)

In [67]:
test_merge_3.head()

event_id     device_id           timestamp  longitude  latitude   city  \
0     2765368  2.973348e+18 2016-05-07 22:52:05  77.225676  28.73014  Delhi   
194   1355198  2.973348e+18 2016-05-01 01:35:45  77.225676  28.73014  Delhi   
388   1630972  2.973348e+18 2016-05-03 00:07:34  77.225676  28.73014  Delhi   
582   2732453  2.973348e+18 2016-05-01 06:15:59  77.225676  28.73014  Delhi   
776   1562542  2.973348e+18 2016-05-01 20:34:19  77.225676  28.73014  Delhi   

     state gender  age   group phone_brand device_model  year  month  \
0    Delhi      M   35  M32-38          优米        UIMI3  2016      5   
194  Delhi      M   35  M32-38          优米        UIMI3  2016      5   
388  Delhi      M   35  M32-38          优米        UIMI3  2016      5   
582  Delhi      M   35  M32-38          优米        UIMI3  2016      5   
776  Delhi      M   35  M32-38          优米        UIMI3  2016      5   

    month_name day_of_week  hour  
0          May    Saturday    22  
194        May      Sunday     1  
388        May     Tuesday     0  
582        May      Sunday     6  
776        May      Sunday    20

In [68]:
test_merge_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16982 entries, 0 to 3640705
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   event_id      16982 non-null  int64         
 1   device_id     16982 non-null  float64       
 2   timestamp     16982 non-null  datetime64[ns]
 3   longitude     16982 non-null  float64       
 4   latitude      16982 non-null  float64       
 5   city          16982 non-null  object        
 6   state         16982 non-null  object        
 7   gender        16982 non-null  object        
 8   age           16982 non-null  int64         
 9   group         16982 non-null  object        
 10  phone_brand   16982 non-null  object        
 11  device_model  16982 non-null  object        
 12  year          16982 non-null  int64         
 13  month         16982 non-null  int64         
 14  month_name    16982 non-null  object        
 15  day_of_week   16982 non-null  obje

In [69]:
test_merge_3.nunique()

event_id        16982
device_id         406
timestamp       16553
longitude         406
latitude          406
city              197
state              26
gender              2
age                51
group              12
phone_brand        26
device_model      191
year                1
month               2
month_name          2
day_of_week         7
hour               24
dtype: int64

## Geospatial Analysis

In [70]:
!pip install folium

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [72]:
import folium

In [73]:
m = folium.Map(location=[28.644800,77.216721])

In [74]:
m

In [79]:
locations_list = test_merge_3[['latitude','longitude']].value_counts().index

In [85]:
locations = locations_list.tolist()

In [90]:
len(locations)

406

In [88]:
mapit = None
# latlon = [ (51.249443914705175, -0.13878830247011467), (51.249443914705175, -0.13878830247011467), (51.249768239976866, -2.8610415615063034)]
mapit = folium.Map( location=[52.667989, -1.464582], zoom_start=6 )
for coord in locations:
    # mapit = folium.Map( location=[ coord[0], coord[1] ] )
    folium.Marker( location=[ coord[0], coord[1] ], fill_color='#43d9de', radius=8 ).add_to( mapit )

mapit

In [91]:
test_merge_3.describe()

event_id     device_id     longitude      latitude           age  \
count  1.698200e+04  1.698200e+04  16982.000000  16982.000000  16982.000000   
mean   1.629002e+06  9.336248e+17     77.586701     21.232838     32.433518   
std    9.368065e+05  1.995877e+18      4.364865      5.495319      9.155786   
min    2.880000e+02  5.036611e+13     69.677548      8.883912     10.000000   
25%    8.204910e+05  1.627458e+16     75.819181     18.506788     26.000000   
50%    1.627304e+06  7.405997e+16     77.078352     22.608837     30.000000   
75%    2.447712e+06  8.524441e+17     80.143569     26.567374     36.000000   
max    3.252680e+06  9.086138e+18     94.190182     33.749656     79.000000   

          year         month          hour  
count  16982.0  16982.000000  16982.000000  
mean    2016.0      4.999941     13.534272  
std        0.0      0.007674      6.242954  
min     2016.0      4.000000      0.000000  
25%     2016.0      5.000000      9.000000  
50%     2016.0      5.000000     14.000000  
75%     2016.0      5.000000     19.000000  
max     2016.0      5.000000     23.000000

In [96]:
list_states = test_merge_3['state'].unique().tolist()

In [101]:
for i in list_states:
 print(i,test_merge_3[test_merge_3['state']==i][['latitude','longitude']].max(),test_merge_3[test_merge_3['state']==i][['latitude','longitude']].min())

Delhi latitude     28.756544
longitude    77.307486
dtype: float64 latitude     28.685333
longitude    77.220866
dtype: float64
Nagaland latitude     25.946122
longitude    94.190182
dtype: float64 latitude     25.728123
longitude    93.824987
dtype: float64
Gujarat latitude     23.969764
longitude    73.139842
dtype: float64 latitude     20.460355
longitude    69.677548
dtype: float64
TamilNadu latitude     13.176833
longitude    80.367855
dtype: float64 latitude      8.883912
longitude    76.997629
dtype: float64
MadhyaPradesh latitude     26.666756
longitude    78.792945
dtype: float64 latitude     21.905980
longitude    74.963796
dtype: float64
Bihar latitude     27.177623
longitude    88.007430
dtype: float64 latitude     24.891077
longitude    84.187234
dtype: float64
AndhraPradesh latitude     19.741445
longitude    83.379038
dtype: float64 latitude     13.255346
longitude    76.934556
dtype: float64
Karnataka latitude     15.299991
longitude    77.671979
dtype: float64 latitude

In [102]:
print(i,test_merge_3[test_merge_3['state']=='Delhi'][['latitude','longitude']].max(),test_merge_3[test_merge_3['state']==i][['latitude','longitude']].min())

Chandigarh latitude     28.756544
longitude    77.307486
dtype: float64 latitude     30.786873
longitude    76.797161
dtype: float64
